### ISSUE TITLE
Automatically cache CompatibilityRule and DisambiguationRule instances

### ISSUE THREAD

**Issue by jvandort** [gradle/gradle#30110]
### Expected Behavior

Instantiating compatibility and disambiguation rules do not affect performance. This is particularly apparent for `CompatibilityRuleChain`, which is called much more often and in loops.

### Current Behavior (optional)

The instantiation of these rules affects performance, as this code is visible in flamegraphs.

### Context

In [InstantiatingAction.java](https://github.com/gradle/gradle/blob/15ada9850698be51581db120376ca0bc6e4b4bc3/subprojects/core/src/main/java/org/gradle/internal/action/InstantiatingAction.java), we cache instantiated compatibility and disambiguation rule instances only if it is marked with `ReuableAction`. That interface has the following javadoc:

```
A marker interface for rules which can be safely reused because they are either
stateless, or effectively immutable. Ideally this should be inferred, which is
why the interface is internal.
```

Note, this is only apparent for non-java projects, as all of Gradle's rules are marked with `ReusableAction`. 

Attached are two flamegraphs which display the performance hit.

android-extra-large-benchmark, 2000 projects:
[androidStudioSync-8 10-20240723143313+0000-cpu-simplified-flames](https://github.com/user-attachments/assets/cd0cda97-fea8-4727-b292-5cfe1575b491)

JFR snapshots from the Block build:
https://github.com/gradle/gradle/issues/29026#issuecomment-2271931626
https://gradle.slack.com/archives/CMBB2G345/p1722970341730999?thread_ts=1721163778.318379&cid=CMBB2G345

**Comment by jvandort** [comment]
We have decided to automatically cache these rules across the board. 

It is a slight behavior change if these rules previously had mutable state or side-effects in their constructor, but we determined this would be a rare scenario, and is behavior we would want to avoid. 

Further, automatic detection of mutability (via reflection) is complex, and not worth it if we expect all rules to be immutable and deterministic anyway.

We will announce this change in the upgrade guide as a potential breaking change.

AGP and KGP rules are verified to be stateless and/or immutable with deterministic constructors
